clippy_lints/src/transmute/mod.rs

```diff
@@ -438,6 +439,36 @@ declare_clippy_lint! {
     "transmute results in a null function pointer, which is undefined behavior"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for slice creation which has larger element before transmute.
+    ///
+    /// ### Why is this bad?
+    /// Creating those slices leads to out-of-bounds read, considered Undefined Behavior.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let i8_slice: &[i8] = &[1i8, 2, 3, 4];
+    //  let i32_slice: &[i32] = unsafe { std::mem::transmute(i8_slice) };
+    /// ```
+    ///
+    /// Use instead:
+    ///
+    /// ```rust
+    /// let i32_slice: &[i32] = i8_slice.iter().map(|item| unsafe { std::mem::transmute(item) }).collect::<Vec<_>>().to_slice();
```

Comment:
Have you tried this suggestion? I don't think this will do what you want: this attempts to transmute `&i8` to `i32`.

Also, isn't the vector you create dropped at the end of the statement? (and don't you mean `.as_slice()`?)

Reply:
Yes, this code does not compile because of the temporary `Vec` being dropped, and the `transmute` on each element is wrong as well.  A "proper" suggestion might be to go through `core::slice::from_raw_parts`, but also because of alignment requirements, I don't think it can be an automated suggestion.  Perhaps "consider `core::slice::from_raw_parts(_mut)` instead" would suffice?